django/template/backends/django.py

```diff
@@ -25,6 +37,46 @@ def __init__(self, params):
         super().__init__(params)
         self.engine = Engine(self.dirs, self.app_dirs, **options)
 
+    def check(self, **kwargs):
+        return [
+            *self._check_string_if_invalid_is_string(),
+            *self._check_for_template_tags_with_the_same_name(),
+        ]
+
+    def _check_string_if_invalid_is_string(self):
+        string_if_invalid = self.engine.string_if_invalid
+        if not isinstance(string_if_invalid, str):
+            error = copy.copy(E002)
+            error.msg = error.msg.format(
+                string_if_invalid, type(string_if_invalid).__name__
+            )
+            return [error]
+        return []
```

Comment:
Whilst we’re touching this, let’s drop the weird `copy()` pattern. Most checks just instantiate the message objects directly, we can do that here too and it makes the code clearer:

```suggestion
        value = self.engine.string_if_invalid
        if not isinstance(value, str):
            return[
                Error(
                    "'string_if_invalid' in TEMPLATES OPTIONS must be a "
                    f"string but got: {value!r} "
                    f"({type(value)}).",
                    obj=self,
                    id="templates.E002",
                )
        return []
```

I swapped for an f-string and used `!r` to repr the value and dropped `__name__` to show the full type path, like:

```
In [1]: class X:
   ...:     pass
   ...:

In [2]: type(X())
Out[2]: __main__.X

In [3]: type(X()).__name__
Out[3]: 'X'
```

I also added the `obj` value for the current backend for identification of the error, now that we are checking the separate the backends.

This will need some repetition of the expected messages in the tests, but that’s in line with other checks and IMO a good thing as it avoids the tests “peeking” into the implementation.

Reply:
Nice that's clearer for sure 🙂